overlord/snapshotstate/snapshotmgr.go

```diff
@@ -0,0 +1,255 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package snapshotstate
+
+import (
+	"encoding/json"
+	"os"
+
+	"gopkg.in/tomb.v2"
+
+	"github.com/snapcore/snapd/client"
+	"github.com/snapcore/snapd/overlord/configstate/config"
+	"github.com/snapcore/snapd/overlord/snapshotstate/backend"
+	"github.com/snapcore/snapd/overlord/snapstate"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/snap"
+)
+
+// SnapshotManager takes snapshots of active snaps
+type SnapshotManager struct {
+	runner *state.TaskRunner
+}
+
+// Manager returns a new SnapshotManager
+func Manager(st *state.State) *SnapshotManager {
+	runner := state.NewTaskRunner(st)
+	runner.AddHandler("save-snapshot", doSave, doForget)
+	runner.AddHandler("forget-snapshot", doForget, nil)
+	runner.AddHandler("check-snapshot", doCheck, nil)
+	runner.AddHandler("restore-snapshot", doRestore, undoRestore)
+	runner.AddCleanup("restore-snapshot", cleanupRestore)
+	return &SnapshotManager{runner: runner}
+}
+
+// KnownTaskKinds is part of the overlord.StateManager interface.
+func (m *SnapshotManager) KnownTaskKinds() []string {
+	return m.runner.KnownTaskKinds()
+}
+
+// Ensure is part of the overlord.StateManager interface.
+func (m *SnapshotManager) Ensure() error {
+	m.runner.Ensure()
+	return nil
+}
+
+// Wait is part of the overlord.StateManager interface.
+func (m *SnapshotManager) Wait() {
+	m.runner.Wait()
+}
+
+// Stop is part of the overlord.StateManager interface.
+func (m *SnapshotManager) Stop() {
+	m.runner.Stop()
+}
+
+type snapshotState struct {
+	SetID    uint64   `json:"set-id"`
+	Snap     string   `json:"snap"`
+	Users    []string `json:"users"`
+	Filename string   `json:"filename"`
+}
+
+func filename(setID uint64, si *snap.Info) string {
+	skel := &client.Snapshot{
+		SetID:    setID,
+		Snap:     si.Name(),
+		Revision: si.Revision,
+		Version:  si.Version,
+	}
+	return backend.Filename(skel)
+}
+
+func prepareSave(task *state.Task) (*snapshotState, *snap.Info, map[string]interface{}, error) {
+	st := task.State()
+	st.Lock()
+	defer st.Unlock()
+
+	var snapshot snapshotState
+	if err := task.Get("snapshot", &snapshot); err != nil {
+		return nil, nil, nil, err
+	}
+	cur, err := snapstate.CurrentInfo(st, snapshot.Snap)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	snapshot.Filename = filename(snapshot.SetID, cur)
+	task.Set("snapshot", &snapshot)
+
+	rawCfg, err := config.GetSnapConfig(st, snapshot.Snap)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	var cfg map[string]interface{}
+	if rawCfg != nil {
+		if err := json.Unmarshal(*rawCfg, &cfg); err != nil {
+			return nil, nil, nil, err
+		}
+	}
+
+	return &snapshot, cur, cfg, nil
+}
+
+func doSave(task *state.Task, tomb *tomb.Tomb) error {
+	snapshot, cur, cfg, err := prepareSave(task)
+	if err != nil {
+		return err
+	}
+	_, err = backend.Save(tomb.Context(nil), snapshot.SetID, cur, cfg, snapshot.Users)
+	return err
```

Comment:
FWIW, looks like we do not need any extra cleanup here. The only 'artifact' created by backend.Save is a zip file which will go away on errors.